st3/lsp_utils/npm_client_handler.py

```diff
@@ -59,51 +92,77 @@ def setup(cls) -> None:
 
-    @property
-    def name(self) -> str:
-        return self.package_name.lower()  # type: ignore
-
-    @property
-    def config(self) -> ClientConfig:
-        assert self.__server
+    @classmethod
+    def config(cls) -> ClientConfig:
+        if cls.__server is None:
+            cls.setup()
 
         configuration = {
             'enabled': True,
-            'command': ['node', self.__server.binary_path] + self.get_binary_arguments(),
+            'command': ['node', cls.__server.binary_path] + cls.get_binary_arguments(),
         }
 
-        configuration.update(self._read_configuration())
-        self.on_client_configuration_ready(configuration)
-        return read_client_config(self.name, configuration)
+        configuration.update(cls._read_configuration())
+        cls.on_client_configuration_ready(configuration)
+        return read_client_config(cls.name, configuration)
+
+    @classmethod
+    def name(cls) -> str:
+        return cls.package_name.lower()  # type: ignore
 
-    def get_binary_arguments(self):
+    @classmethod
+    def needs_update_or_installation(cls) -> bool:
+        return False
+
+    @classmethod
+    def install_or_update(cls) -> None:
+        pass
+
+    @classmethod
+    def standard_configuration(cls) -> ClientConfig:
+        return cls.config()
+
+    @classmethod
+    def adjust_configuration(cls, configuration: ClientConfig) -> ClientConfig:
+        return configuration
+
+    @classmethod
+    def can_start(cls, window: sublime.Window, initiating_view: sublime.View,
+                  workspace_folders: List[WorkspaceFolder], configuration: ClientConfig) -> Optional[str]:
+        return None
+
+    @classmethod
+    def get_binary_arguments(cls):
         """
         Returns a list of extra arguments to append when starting server.
         """
         return ['--stdio']
 
-    def _read_configuration(self) -> Dict:
+    @classmethod
+    def _read_configuration(cls) -> Dict:
         settings = {}  # type: Dict
-        loaded_settings = sublime.load_settings(self.settings_filename)  # type: sublime.Settings
+        loaded_settings = sublime.load_settings(cls.settings_filename)  # type: sublime.Settings
 
         if loaded_settings:
-            migrated = self._migrate_obsolete_settings(loaded_settings)
-            changed = self.on_settings_read(loaded_settings)
+            migrated = cls._migrate_obsolete_settings(loaded_settings)
+            changed = cls.on_settings_read(loaded_settings)
             if migrated or changed:
-                sublime.save_settings(self.settings_filename)
+                sublime.save_settings(cls.settings_filename)
 
             for key, default in CLIENT_SETTING_KEYS.items():
                 settings[key] = loaded_settings.get(key, default)
 
         return settings
 
-    def on_settings_read(self, settings: sublime.Settings):
+    @classmethod
+    def on_settings_read(cls, settings: sublime.Settings):
```

Comment:
This is not backwards compatible, is it? Previously it was a method of the instance, now it's a method of the class. The overrides won't work.

Reply:
You can call classmethods on an instance just fine :)

```
class Foo:

    @classmethod
    def f(cls):
        print("hello")


foo = Foo()
foo.f()
```

Reply:
Also @classmethods can be overridden. @staticmethod can not be overridden. @classmethod is magic.

Reply:
But this (which is how it would have to be done today) doesn't work:
```py
class A(object):
    @classmethod
    def configuration(cls):
        cls.on_read_configuration()

    @classmethod
    def on_read_configuration(cls):
        return 'Called on_read_configuration'


class B(A):
    def on_read_configuration(self):
        print('Ran overriden on_read_configuration')


b = B()
b.configuration()
```

Reply:
The only method that is in use like that is `on_client_configuration_ready`. It is used by LSP-json, LSP-vue and LSP-intelephense.

LSP-vue and LSP-intelephense don't reference instance properties at all, so it's okay to use a hack like this:

```
        called = False
        for c in cls.__subclasses__():
            f = getattr(c, 'on_client_configuration_ready', None)
            if callable(f):
                f(cls, configuration)
                called = True
                break
        if not called:
            cls.on_client_configuration_ready(configuration)
```

For LSP-json we can't do that as it accesses the instance property `self._default_schemas`.

IMO it's okay to rewrite that a bit to make it a classmethod:

```python
class LspJSONPlugin(NpmClientHandler):
    package_name = __package__
    server_directory = 'vscode-json-languageserver'
    server_binary_path = os.path.join(
        server_directory, 'node_modules', 'vscode-json-languageserver', 'bin', 'vscode-json-languageserver'
    )
    _default_schemas = []  # type: List[Dict]

    @classmethod
    def on_client_configuration_ready(cls, configuration: Dict):
        if not cls._default_schemas:
            schemas = ['schemas_extra.json', 'schemas.json']
            for schema in schemas:
                path = 'Packages/{}/{}'.format(cls.package_name, schema)
                cls._default_schemas.extend(json.loads(ResourcePath(path).read_text()))

        configuration['settings'].setdefault('json', {})['schemas'] = cls._default_schemas
```

Reply:
I guess this should work https://github.com/sublimelsp/lsp_utils/pull/9